### PR TITLE
Mm nk join existing list

### LINKED ISSUE #6: 5. As a user, I want to join an existing shopping list so I can share a shopping list with someone
Tokens can be shared with other users to allow them to co-manage a given list.

Joining an existing list consists of the following:

* User enters an existing token
* Save the token to `localStorage`
* Show the user the existing list

AC:

* When the user does not already have a token in `localStorage`, on the onboarding/home screen, the user can enter an existing token
* Entering the token and hitting submit saves the token to `localStorage`, effectively giving them joint control of the list
* On submit, show an error if the token does not exist
* If they get an error message, allow them to try again or create a new list

### PR DESCRIPTION
**Pull request by nabilkaz**
## Description
This PR adds the functionality that allows a person to join an existing shopping list. It does this by adding a `ShareList` module to handle a person joining a list.

The component handles the following processes:
1. Clicking "Join existing shopping list" should take us to `./join-existing`
2. `./join-existing` we should have the capacity to enter a share code
3. When share is submitted we need the capacity to validate share code
    - if share code is valid we need to:
        -  Set token of list to `localStorage`
        - Route to `./list-view`
    - If share is invalid we pop up an alert window

Additionally the route `.join-existing` was added to our app along with an updated `href` property to the "join an existing shopping list" link found at route `/`. 

### Collection Validation & Routing logic
The component `ShareList` in  `Components/shareList.js`  contains the routing logic within the method `getExistingList` on [line 17](https://github.com/the-collab-lab/tcl-7-smart-shopping-list/blob/d787c3988824a193841f672f4f63070ae06372e4/src/Components/shareList.js#L17). 
1. We create an interface with our Cloud Firestore service and set it to the variable `db`
2. We take the value from the input value that is stored in the components state `this.state.token` pass it as the collection path to the [method `db.collection`](https://firebase.google.com/docs/reference/js/firebase.firestore.Firestore#collection) which gets a [`CollectionReference` instance](https://firebase.google.com/docs/reference/js/firebase.firestore.CollectionReference) based on this path. We then chain a [`.get` method](https://firebase.google.com/docs/reference/js/firebase.firestore.CollectionReference#get) which query's the database and returns a [promise-like object `QuerySnapshot`](https://firebase.google.com/docs/reference/js/firebase.firestore.QuerySnapshot). This is stored to the constant `promise` on [line 19](https://github.com/the-collab-lab/tcl-7-smart-shopping-list/blob/d787c3988824a193841f672f4f63070ae06372e4/src/Components/shareList.js#L19). 😳well that was a lot...
2.  Now we're working asynchronously, we chain a `.then` method which returns a `QuerySnapshot` object which we pass as an argument to the callback function passed to the `.then` method. 
3. In the callback function we house the pass-fail logic triggered by the return value of `!querySnapshot.empty` 
    - If the QuerySnapshot object is empty (no documents) we know it doesn't exist and show an alert.
   - Otherwise we will call `setLocalToken` and pass `this.state.token` in as the token to set in `localStorage`. We then route the client to `./list-view` which uses the `token` stored in `localStorage` to retrieve and serve the list.

The event handler `handleCheckShareCode` is registered to the `onClick` event of the "Join Shopping List" button.

`handleCheckShareCode` prevents the default behavior of the form and performs the runs `getExistingList`.

### Styling of `./join-existing`
We copied the existing style and structure of `SignIn` and added the Google Material style-like inputs that hover on focus. 

Loom demo: https://www.loom.com/share/89c426933b0147ff823377c5321869a4

Hover input:
image

### Accessibility
We passed all tests and fulfilled the tabbing requirements.
 image

## Related Issue
Closes #6 

## Acceptance Criteria

- When the user does not already have a token in localStorage, on the onboarding/home screen, the user can enter an existing token
- Entering the token and hitting submit saves the token to localStorage, effectively giving them joint control of the list
- On submit, show an error if the token does not exist
- If they get an error message, allow them to try again or create a new list

## Type of Changes

|     | Type                       |
| --- | -------------------------- |
|    | :bug: Bug fix              |
|  ✓ | :sparkles: New feature     |
|  ✓  | :hammer: Refactoring       |
|    | :100: Add tests            |
|    | :link: Update dependencies |
|    | :scroll: Docs              |

### Before
`./joing-existing` did not exist

### After
Loom demo: https://www.loom.com/share/89c426933b0147ff823377c5321869a4

image

## Testing Steps / QA Criteria

Setup:
1. On your localhost in chrome open the developers tools
2. Select the “Application” tab in the console's top menu.
3. Select “Local Storage” in the console's left menu.
4. Right-click and click clear to delete the local storage. image
5. In your browser navigate to root `/`
6. you should see image

Test 1:
1. On root `/` click `join existing shopping list` 
image
2. If navigates to `./join-existing` then passed

Test 2: 
1. On `./join-existing` click `share code` and enter `not a real token` 
image
2. Click `Join shopping list` 
image
3. If alert pops up then passed 
image

Test 3: 
1. On `./join-existing` click `share code` and enter `user`
image
2. Click `Join shopping list`
3. Passes if:
    - Navigates to `./list-view` 
image
    - Entering `window.localStorage.getItem('token') === 'user'` into the console returns `true`. 
image

### CODE REVIEW

File: src/App.js
Comment (alejandronanez) [inline src/App.js]: In the future (let's keep this as is because it's working fine!), let's try a couple of things:

1. Remove the `onClick` event from the `button` element, and instead add it to the `form` element as a `onSubmit` event. As we talked in the first TCL call, thanks to @msholty-fd!, adding the event to the `onSubmit` will give your users a better experience because every time they hit `enter` the form will be submitted. Right now, the user will have to either click or tab + space the button to get this working.
Again, we don't need to change this as I believe it's out of scope and was already introduced in a previous PRs

2. I'm not really sure about the usage of `action="./add-item"` in the form tag, what this is doing is that it's redirecting right after sending the form (without any validation) and it always asumes the happy path. You have two options:
a. You can use the [useHistory hook](https://reacttraining.com/react-router/web/api/Hooks/usehistory), but that would mean that you have to replace the your class component with a functional component
b. You can use [withRouter](https://reacttraining.com/react-router/web/api/withRouter) to get `history` via `this.props.history`
Once you're done with that, you can use the `history.push` method to navigate to the next page, `history.push("/home");`

I'd like us to change this before merging **if you're 👍 with it and have time to implement it**, if not, I'd suggest moving forward as this is working right now!

File: src/Components/shareList.js
Comment (alejandronanez) [inline src/Components/shareList.js]: So glad you brought up this pattern 👀.

You can get the same result by writing:

```js
class ShareList extends React.Component {

  constructor(props) {
    super(props)
    this.state = { token: ''};
  }

  updateInput = (event) => {
    this.setState({....})
  }

  // More code omitted for brevity
}
```

And you can go even further:

```js
class ShareList extends React.Component {

  state = { token: '' }

  updateInput = (event) => {
    this.setState({....})
  }

  // More code omitted for brevity
}
```

File: src/Components/shareList.js
Comment (alejandronanez) [inline src/Components/shareList.js]: I suggest using React Router's `history` API. I already mentioned this in my comment in `shareList` component.

- https://reacttraining.com/react-router/web/api/withRouter
- https://reacttraining.com/react-router/web/api/Hooks/usehistory

Comment (alejandronanez) [reply src/Components/shareList.js]: Not a deal breaker though, if you're short on time, leave this as is!

File: src/lib/token.js
Comment (alejandronanez) [inline src/lib/token.js]: I believe we can simplify this to a one liner if by doing

```js
window.localStorage.setItem('token', token || getToken());
```

I don't see much value of moving `token` and `window.localStorage` into their own variables  just to use them in the next line. 😄

**Review by alejandronanez — APPROVED** [review]
Great work team, love to see this working so nicely :) 🎉

I left some recommendations that are not mandatory, so take them into account for the next time around.

I strongly suggest you to start peeking at functional components and hooks, there's nothing wrong with using classes but all the new components you're going to build in a real product or the libraries that you're going to use are migrating to `hooks` and you won't be able to use them as you're still using classes. Again, this is not mandatory but it a good thing to start practicing! :)

File: src/App.css
Comment (msholty-fd) [inline src/App.css:46]: these kind of selectors are scary, since they won't get hashed by CSS Modules, and affect so many elements regardless of page or component. I hope you intend for these to be site-wide!

File: src/App.js
Comment (msholty-fd) [inline src/App.js]: Like @alejandronanez said, you should get used to using functional components. It's a bit strange to me to see a component go from function _to_ class based here. If you are looking to write event handlers in the component, you can do so like this:
```
function SignIn() {
  function handleClickSetLocalToken() {
  }

  return < ... />
}
```

File: src/App.js
Comment (msholty-fd) [reply src/App.js]: I really agree with point 2 here @alejandronanez -

I believe having an `onSubmit` handler is a good idea, but the `action` attribute makes this form behave as if we weren't working within the confines of a JS application. The `action` attribute will `POST` the form's data to that url, but since we are in a JS application we can validate it immediately here.

File: src/Components/shareList.js
Comment (msholty-fd) [reply src/Components/shareList.js]: Part of the reason why I like to stray away from class based components is the fact that we have to use `this`, and also manage `this` like such. I'd prefer to just to use JavaScript closures as intended with functional components, and not worry about `this` ever!

```
function ShareList() {
  const [token, setToken] = useState();

  function updateInput(e) {
    setToken(e.currentTarget.value)
  }

  return <input onChange={updateInput}>
}
```

File: src/Components/shareList.js
Comment (msholty-fd) [reply src/Components/shareList.js]: One note about `useHistory` - it can't be used in this class based component. You'd have to wrap this component with `withRouter` (a Higher order Component, or HoC)

File: src/App.js
Comment (msholty-fd) [inline src/App.js]: Because this is a vanilla `<a>` tag and not a [`<Link>` component](https://reacttraining.com/react-router/web/api/Link), this redirect is a hard redirect that loads the app again, and then we run into [the netlify redirect issue](https://the-collab-lab.slack.com/archives/C012AAC9D9S/p1590277044010300) that we haven't resolved yet. I'm having trouble getting to this page as a result.

Before merging, could you change this to `<Link>` instead of `<a>`?

**Review by msholty-fd — APPROVED** [review]
I tested this with y'all over a Zoom call, so I know the page works, but I am just having issues visiting it in the netlify demo because of the reason I mentioned in a comment. Approving now, but make sure to take a look into that before merging 😄

File: src/lib/token.js
Comment (DeeThompson) [reply src/lib/token.js]: Hi Nabil and Megan, I ran the project on my localhost, checked through everything and it works the way it's stated in the AC, good job!

File: src/App.js
Comment (nabilkaz) [reply src/App.js]: Thanks so much! Spent the better part of this morning familiarizing myself with Hooks 💪🏽

I changed the `SignIn` class component to a functional one, imported `useHistory` from `react-router-dom` and added an `onSubmit` event handler to perform both setting a token in local storage and routing to `./add-item`.

Additionally, I wrote up a short summary of what I learned:
> Before React Hooks the biggest difference between a class and functional component was syntax, state, and handling lifecycle. If you wanted to create a dynamic component, one that used state and lifecycle to change states, you’d use a class component. Functional components could only use state if you lifted state up to the parent component and passed it down to the functional component via props.
> 
> The choice was clear: Class components to take advantage of state and lifecycle methods. Otherwise, use functional components to take advantage of their clear and readable syntax and performance benefits.
> 
> This left us with three points of frustration: 
> 1. It was hard to reuse stateful logic between components
> 2. Complex components became hard to understand
> 3. Classes confused both people and machines
> 
> React 16.8 introduced Hooks which enabled functional components to use state and lifecycle events. This meant you could advantage of the clear syntax and performance of functional components and the state and lifecycle methods that were once only available for class components.

File: src/App.css
Comment (nabilkaz) [reply src/App.css:46]: I do! These are resets but I think you're right. This could be done with much more intent. Importing normalize.css and tailwind might be a better option than the route I've been going. I also just found out we don't have to match the example app so I might redesign the visuals for the app 🎉

**Pull request merged**
